spinalcordtoolbox/deepseg/inference.py

```diff
@@ -77,3 +91,158 @@ def segment_and_average_volumes(model_paths, input_filenames, options):
+
+    Author: Naga Karthik
+    Original script: https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/blob/e65478099d026f865b7f1d7d0082e6e9a507a744/monai/run_inference_single_image.py
+    """
+    # Copy the file to the temporary directory using shutil.copyfile
+    path_img_tmp = os.path.join(tmpdir, os.path.basename(path_img))
+    shutil.copyfile(path_img, path_img_tmp)
+    logger.info(f'Copied {path_img} to {path_img_tmp}')
+
+    # define inference patch size and center crop size
+    crop_size = (64, 192, -1)
+    inference_roi_size = (64, 192, 320)
+
+    # define the dataset and dataloader
+    test_loader, test_post_pred = ds_monai.prepare_data(path_img_tmp, tmpdir, crop_size=crop_size)
+    batch = next(iter(test_loader))
+
+    # Run nnUNet prediction
+    print('Starting inference...')
+    start = time.time()
+
+    # run inference
+    test_input = batch["image"].to(torch.device("cpu"))
+    batch["pred"] = sliding_window_inference(test_input, inference_roi_size, mode="gaussian",
+                                             sw_batch_size=4, predictor=predictor, overlap=0.5, progress=False)
+    pred = ds_monai.postprocessing(batch, test_post_pred)
+
+    end = time.time()
+    print('Inference done.')
+    total_time = end - start
+    print(f'Total inference time: {int(total_time // 60)} minute(s) {int(round(total_time % 60))} seconds')
+
+    # this takes about 0.25s on average on a CPU
+    # image saver class
+    _, fname, ext = extract_fname(path_img)
+    postfix = "seg"
+    target = f"_{postfix}"
+    pred_saver = SaveImage(
+        output_dir=tmpdir, output_postfix=postfix, output_ext=ext,
+        separate_folder=False, print_log=False)
+    # save the prediction
+    fname_out = os.path.join(tmpdir, f"{fname}_{postfix}{ext}")
+    logger.info(f"Saving results to: {fname_out}")
+    pred_saver(pred)
+
+    return [fname_out], [target]
+
+
+def segment_nnunet(path_img, tmpdir, predictor):
+    """
+    This script is used to run inference on a single subject using a nnUNetV2 model.
+
+    Author: Jan Valosek, Naga Karthik
+    Original script: https://github.com/ivadomed/model_seg_sci/blob/4184bc22ef7317b3de5f85dee28449d6f381c984/packaging/run_inference_single_subject.py
+    """
+    # Copy the file to the temporary directory using shutil.copyfile
+    path_img_tmp = os.path.join(tmpdir, os.path.basename(path_img))
+    shutil.copyfile(path_img, path_img_tmp)
+    logger.info(f'Copied {path_img} to {path_img_tmp}')
+
+    # Get the original orientation of the image, for example LPI
+    orig_orientation = get_orientation(Image(path_img_tmp))
+
+    # Get the orientation used by the model
+    if "SCI" in predictor.plans_manager.dataset_name:
+        model_orientation = "RPI"
+    else:
+        assert "levels" in predictor.plans_manager.dataset_name  # rootlets dataset
+        model_orientation = "LPI"
```

Comment:
Right now, the two `nnunetv2` models are _very_ similar in interface, with the only differences being:

1. The orientation expected by the model.
2. The number of output images (SCI = multiclass, lesion+SC, Rootlets = single output)

The way I've handled this is by "special case"-ing the differing logic by checking against the model. But, I'm not 100% sure how I feel about this, so input would be much appreciated. :)

Reply:
correct me if i'm wrong -- I remember discussing in the SCT meeting about having SCT's function call decide what model to use. For e.g. `sct_segment_sci` would call the SCI model and `sct_segment_rootlets` would call the rootlets model. If that's the case, then we don't have to rely on the dataset name right?

Reply:
<details>
<summary> Off-topic response about CLI command names </summary>

You're right! We absolutely discussed this in the [2023-11-30](https://docs.google.com/document/d/1ItApJQfajO2lRzOU2yenWbeRg6alfsdut3J4AVVdo78/edit#heading=h.sl17y9esapii) meeting as proposed by Jan. But, I think the conclusion of that meeting was to stick to `sct_deepseg`?

Still, the interface design is an ongoing discussion point, since it was brought up in the most recent SCT meeting ([2024-01-11](https://docs.google.com/document/d/1ItApJQfajO2lRzOU2yenWbeRg6alfsdut3J4AVVdo78/edit#heading=h.7phkpm9qz4ll)) as well. That said, I think that the best place for follow-up discussion would be https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4318.

Right now, this PR sticks to `sct_deepseg` CLI command.

</details>

It sounds like you're describing the user-facing "CLI level". But, even if we were to change the "CLI level" design by using dedicated commands for each model (`sct_segment_sci`, `sct_segment_rootlets`), then the two models would probably still share code on the "API level", since they were both developed using similar `nnunetv2` workflows.

So, no matter what the CLI looks like, we would still need to figure out how to pass model information from the CLI level to the shared `segment_nnunet` API level function.

Reply:
Thanks for clarifying! I definitely mixed up CLI- and API-levels in my head when I wrote this.